### PR TITLE
Diff fragment

### PR DESCRIPTION
**Pull request by mlrawlings**
## Description

**Introducing HTMLFragment**

- Updates the diffing algorithm to use an HTMLFragment node as an abstraction rather than keeping track of startNode and endNode all throughout the diffing algorithm.
- Uses the HTMLFragment for the `<${dynamic}>` tag and `<include>` tags to preserve server-rendered content for which the `renderBody` is not available in the browser.

**Changes to keys**

- Component ids are based on the resulting parent tree (not the owner tree).  This means we cannot rely on the ids in the global lookup, so component key/refs are now also stored on the component instance.
- Autokeyed elements are now stored on the parent rather than the owner.  User assigned key/refs are still stored on the owner component.  Because of this, user assigned keys are now prefixed to differentiate them from autokeys.  This also has the benefit that assigning numeric keys can no longer conflict with the autokeys.
- Static node trees are now only auto assigned a key for the top-level node (instead of all nodes).  This is because:
  - When updating, only the top-level node is considered.  
  - When static nodes are hoisted out of the render method, they are not associated with the owner component

**Changes to Hydration**

- Server comment starting markers now have the component's key serialized so the component can be attached to its owner
- Server comment markers no longer have the id on the closing marker, it is stack based.
- Normalize differences between hydration and client-rendering, so post mount the DOM looks the same regardless of whether a component was server or client rendered.

## Motivation and Context

- Fixes https://github.com/marko-js/marko/issues/1059
[test/components-browser/fixtures/force-update-dom-exception-3/test.js](https://github.com/marko-js/marko/pull/1094/files#diff-4b4c3fb579dc70a5d1a606c1d21b0bb5)
- Fixes https://github.com/marko-js/marko/issues/1051
[test/components-browser/fixtures/nested-fragments/test.js](https://github.com/marko-js/marko/pull/1094/files#diff-71dfc9f3f89d8b9ba2f4c814bfa2b2da)
- Fixes https://github.com/marko-js/marko/issues/1033
[test/components-browser/fixtures/top-level-transcluded-content/test.js](https://github.com/marko-js/marko/pull/1094/files#diff-a8a998faf0222cf3e7c89b12a627dddc)
- Fixes https://github.com/marko-js/marko/issues/924 
(Nodes are no longer removed before reinserting, future mutation tests will verify)
- Fixes https://github.com/marko-js/marko/issues/920 
(I think... we have not been able to reproduce)
- Fixes https://github.com/marko-js/marko/issues/912
[test/components-pages/fixtures/preserve-transcluded-content-on-state-change/tests.js](https://github.com/marko-js/marko/pull/1094/files#diff-4c548340c40b7892eae861fb3d26c902)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] I have updated/added documentation affected by my changes.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

_Disclaimer: Contributions via GitHub pull requests are gladly accepted from their original author. Along with any pull requests, please state that the contribution is your original work and that you license the work to the project under the project's open source license. Whether or not you state this explicitly, by submitting any copyrighted material via pull request, email, or other means you agree to license the material under the project's open source license and warrant that you have the legal authority to do so._

### CODE REVIEW

**Comment by cameronbraid** [comment]
Great work..

This looks promising.  I will try merging it into my fork and give it a go

Thanks Cameron

**Comment by cameronbraid** [comment]
One issue so far :

```
TypeError
    at VElement.___appendChild (VNode.js?08f2:58)
    at AsyncVDOMBuilder.___beginNode (AsyncVDOMBuilder.js?14fc:71)
    at AsyncVDOMBuilder.___beginFragment (AsyncVDOMBuilder.js?14fc:214)
    at doInclude (include-tag.js?5030:42)
    at includeTag (include-tag.js?5030:56)
    at wrappedRenderer (helpers.js?a540:195)
    at render (dn-textarea.marko?7457:52)
    at renderFunc (renderer.js?3aae:225)
    at renderer (renderer.js?3aae:237)
```

**Comment by cameronbraid** [comment]
I was doing an include tag in a textarea like

```
<textarea><include(input.renderBody)/></textarea>
```

I refactored to use 

```
<textarea>${input.value}</textarea>
```

and all is good ;)

**Comment by cameronbraid** [comment]
I have run my chaos test generator that I used for 1059 at https://github.com/cameronbraid/marko-dom-issue and there are no issues !

Well done.

Thank you, Thank you, Thank you/

File: src/components/Component.js [hunk 1]
```diff
@@ -137,12 +137,6 @@ function checkInputChanged(existingComponent, oldInput, newInput) {
     return false;
 }
 
-function getNodes(component) {
-    var nodes = [];
-    component.___forEachNode(nodes.push.bind(nodes));
-    return nodes;
-}
-
```
Comment (mdalhaz) [inline src/components/Component.js]: now i beed to do that

**Comment by cameronbraid** [comment]
Sad to say I have run into a dom exception using this patch :

```
          Uncaught TypeError: Cannot read property 'insertInto' of null
            at insertBefore (helpers.js?2a75:2)
            at morphChildren (index.js?8192:187)
            at morphEl (index.js?8192:636)
            at morphChildren (index.js?8192:235)
            at morphComponent (index.js?8192:94)
            at morphChildren (index.js?8192:198)
            at morphEl (index.js?8192:636)
            at morphChildren (index.js?8192:235)
            at morphEl (index.js?8192:636)
            at morphChildren (index.js?8192:235)
            at morphdom (index.js?8192:645)
            at eval (Component.js?2cba:511)
            at Object.batchUpdate [as ___batchUpdate] (update-manager.js?8c0e:63)
            at Component.___rerender (Component.js?2cba:494)
            at Component.update (Component.js?2cba:448)
            at updateComponents (update-manager.js?8c0e:44)
            at updateUnbatchedComponents (update-manager.js?8c0e:16)
            at eval (nextTick-browser.js?bdf1:18)
```

I havn't yet had the chance to boil it down to a stand alone repo

**Comment by cameronbraid** [comment]
Some good news, It appears that it was my code that uses the marko compiler that was the source cause of this.   It was, in a particular case, using a duplicate templatePath parameter (one used for another component)

**Comment by scttdavs** [comment]
just want to add, we fall into this bucket:

> `<include>` tags to preserve server-rendered content for which the renderBody is not available in the browser.

And it fixes it for us!

**Pull request merged**